README.md

```diff
@@ -1,26 +1,643 @@
+  invoke a transition.
+* _Transition_ is an operation that receive state and return new state.
+
+To add microstates to your node project,
+
+```bash
+npm install --save microstates
+
+# yarn add microstates
+```
+
+`microstates` module exports a microstate constructor as `default` and built in types: `Number`,
+`String`, `Boolean`, `Object` and `Array`.
+
+_Note_: We recommend `import * as MS from 'microstates'` syntax, to hide built-in types behind a
+namespace. Otherwise they'll overwrite built in JavaScript classes in the module.
+
+Let's see some code,
+
+```js
+import microstate, * as MS from 'microstates';
+
+microstate(MS.Number, 42).increment().state;
+//=> 43
+
+class MyCounter {
+  count = MS.Number;
+}
+
+// creating microstate from composed states
+microstate(MyCounter).state;
+// => { count: 0 }
+
+microstate(MyCounter).count.increment().state;
+// => { count: 1 }
+
+class MyModal {
+  isOpen = MS.Boolean;
+  title = MS.String;
+}
+
+// you can restore a microstate from previous value
+microstate(MyModal, { isOpen: true, title: 'Hello World' }).state;
+// => { isOpen: true, title: 'Hello World' }
+
+microstate(MyModal, { isOpen: true, title: 'Hello World' }).isOpen.toggle().state;
+// => { isOpen: false, title: 'Hello World' }
+
+// lets composed multiple state together
+class MyState {
+  modal = MyModal;
+  counter = MyCounter;
+}
+
+microstate(MyState).state;
+// => {
+//  modal: { isOpen: false, title: '' },
+//  counter: { count: 0 }
+// }
+
+microstate(MyState)
+  .counter.increment()
+  .modal.title.set('Hello World')
+  .modal.isOpen.set(true).state;
+// => {
+//  modal: { isOpen: true, title: 'Hello World' },
+//  counter: { count: 1 }
+// }
+```
+
+# microstate constructor
+
+Microstate constructor creates a microstate. It accepts two arguments: `Type` class that describes
+the structure of your state and `value` that is the initial state.
+
+```js
+// initial value is undefined so state will be default value of String which is an empty string
+microstate(MS.String).state;
+// => ''
+
+// initial value is 'hello world' so state will be hello world.
+microstate(MS.String, 'hello world').state;
+// => ''
+```
+
+The object returned from the constructor has all of the transitions for the structure.
+
+```js
+let ms = microstate(MS.Object);
+
+console.log(ms);
+//  {
+//    assign: Function
+//    set: Function
+//  }
+
+ms.assign({ hello: 'world', hi: 'there' }).state;
+// => { hello: 'world', hi: 'there' }
+```
+
+# Composition
+
+Microstates are designed to be composable. You can define custom microstate types and nested them
+necessary. Microstates will take care of ensuring that all transitioning composed microstates is
+done immutably.
+
+Defining custom types is done using ES2016 classes syntax. In JavaScript classes are functions. So
+any function can work as a type, but we recommend using class syntax.
+
+```js
+microstate(class {});
+// => {
+//      merge: Function
```

Comment:
I don't see merge function when I tried this in repl. Is this a WIP?

Reply:
No, this is a mistake in the documentation. Merge function will only appear when a type has nested states.